File: skills/skill-audit-ts/scripts/validate_skill.ts
#!/usr/bin/env -S npx tsx
/**
 * Validate an agent skill against the Agent Skills specification.
 *
 * Usage:
 *   validate_skill.ts <skill-directory> [--json] [--exit-on-warn]
 *
 * Default output is one issue per line, prefixed with FAIL: or WARN: and a
 * machine-readable code in brackets. With --json, emits a structured object
 * on stdout.
 *
 * Exit codes:
 *   0   no FAIL-level issues (and no WARN if --exit-on-warn)
 *   1   one or more FAIL-level issues
 *   2   bad invocation (path missing, not a directory, etc.)
 */

import { existsSync, readFileSync, statSync } from "node:fs";
import { resolve, relative, join } from "node:path";
import { parseArgs } from "node:util";
import { emitError, parseFrontmatter, sanitizeForEcho, type Frontmatter } from "./skill_lib.js";

const NAME_RE = /^[a-z0-9]+(-[a-z0-9]+)*$/;
const MD_LINK_RE = /\]\(([^)]+\.md)\)/g;
const SCRIPT_REF_RE = /`(scripts\/[^`\s]+\.(?:py|sh|js|ts))`/g;
const MAX_NAME = 64;
const MAX_DESCRIPTION = 1024;
const MIN_DESCRIPTION_WARN = 60;
const MAX_COMPATIBILITY = 500;

const ALLOWED_KEYS = new Set([
  // agentskills.io base spec
  "name",
  "description",
  "license",
  "compatibility",
  "metadata",
  "allowed-tools",
  // Platform extensions (Claude Code, Copilot, Codex, VS Code)
  "when_to_use",
  "argument-hint",
  "arguments",
  "disable-model-invocation",
  "user-invocable",
  "model",
  "effort",
  "context",
  "agent",
  "hooks",
  "paths",
  "shell",
]);

export interface Issue {
  severity: "fail" | "warn";
  code: string;
  message: string;
  field?: string;
}

function issueLine(issue: Issue): string {
  const prefix = issue.severity === "fail" ? "FAIL" : "WARN";
  return `${prefix}: [${issue.code}] ${issue.message}`;
}

export function validate(skillDir: string): Issue[] {
  const issues: Issue[] = [];
  const skillMd = join(skillDir, "SKILL.md");
  if (!existsSync(skillMd)) {
    return [
      {
        severity: "fail",
        code: "validate.skill-md.missing",
        message: `SKILL.md not found at ${sanitizeForEcho(skillMd)}`,
      },
    ];
  }

  const text = readFileSync(skillMd, "utf-8");
  const [fm, body] = parseFrontmatter(text);
  if (fm === null) {
    return [
      {
        severity: "fail",
        code: "validate.frontmatter.missing",
        message: "SKILL.md is missing YAML frontmatter (must start with '---')",
      },
    ];
  }

  checkName(fm, skillDir, issues);
  checkDescription(fm, issues);
  checkCompatibility(fm, issues);
  checkUnknownKeys(fm, issues);
  checkReferences(skillDir, body, issues);
  return issues;
}

function checkName(fm: Frontmatter, skillDir: string, issues: Issue[]): void {
  const name = typeof fm.name === "string" ? fm.name : "";
  if (!name) {
    issues.push({
      severity: "fail",
      code: "validate.name.missing",
      message: "'name' field is required",
      field: "name",
    });
    return;
  }
  const safe = sanitizeForEcho(name, 80);
  if (name.length > MAX_NAME) {
    issues.push({
      severity: "fail",
      code: "validate.name.too-long",
      message: `'name' is ${name.length} chars (max ${MAX_NAME})`,
      field: "name",
    });
  }
  if (!NAME_RE.test(name)) {
    issues.push({
      severity: "fail",
      code: "validate.name.bad-format",
      message:
        "'name' must be lowercase a-z/0-9 with single hyphens, no leading/trailing/" +
        `consecutive hyphens. Got: '${safe}'`,
      field: "name",
    });
  }
  const dirName = skillDir.split(/[/\\]/).filter(Boolean).pop() ?? "";
  if (name !== dirName) {
    issues.push({
      severity: "fail",
      code: "validate.name.dir-mismatch",
      message: `'name' ('${safe}') must match parent directory ('${sanitizeForEcho(dirName)}')`,
      field: "name",
    });
  }
}

function checkDescription(fm: Frontmatter, issues: Issue[]): void {
  const desc = typeof fm.description === "string" ? fm.description : "";
  if (!desc) {
    issues.push({
      severity: "fail",
      code: "validate.description.missing",
      message: "'description' is required and must be non-empty",
      field: "description",
    });
    return;
  }
  if (desc.length > MAX_DESCRIPTION) {
    issues.push({
      severity: "fail",
      code: "validate.description.too-long",
      message: `'description' is ${desc.length} chars (max ${MAX_DESCRIPTION})`,
      field: "description",
    });
  }
  if (desc.length < MIN_DESCRIPTION_WARN) {
    issues.push({
      severity: "warn",
      code: "validate.description.too-short",
      message:
        `'description' is only ${desc.length} chars — likely too short to convey ` +
        "when to trigger",
      field: "description",
    });
  }
}

function checkCompatibility(fm: Frontmatter, issues: Issue[]): void {
  const compat = fm.compatibility;
  if (typeof compat === "string" && compat && compat.length > MAX_COMPATIBILITY) {
    issues.push({
      severity: "fail",
      code: "validate.compatibility.too-long",
      message: `'compatibility' is ${compat.length} chars (max ${MAX_COMPATIBILITY})`,
      field: "compatibility",
    });
  }
}

const ALLOWED_KEYS_LIST = [...ALLOWED_KEYS].sort().join(", ");

function checkUnknownKeys(fm: Frontmatter, issues: Issue[]): void {
  const unknown = Object.keys(fm)
    .filter((k) => !ALLOWED_KEYS.has(k))
    .sort();
  for (const key of unknown) {
    const safe = sanitizeForEcho(key, 64);
    issues.push({
      severity: "warn",
      code: "validate.frontmatter.unknown-key",
      message: `unknown frontmatter key '${safe}' (spec defines: ${ALLOWED_KEYS_LIST})`,
      field: key,
    });
  }
}

function withinSkill(skillRoot: string, target: string): boolean {
  const rel = relative(skillRoot, target);
  return rel === "" || (!rel.startsWith("..") && !rel.startsWith(`..${"/"}`) && rel !== "..");
}

function checkRefsExist(
  skillDir: string,
  skillRoot: string,
  refs: string[],
  code: string,
  label: string,
  issues: Issue[],
): void {
  for (const ref of refs) {
    // Skip template placeholder paths like `references/<topic>.md`.
    if (/[<>${}]/.test(ref)) continue;
    const target = resolve(skillDir, ref);
    if (!withinSkill(skillRoot, target)) continue;
    if (!existsSync(target)) {
      issues.push({
        severity: "warn",
        code,
        message: `referenced ${label} does not exist: ${sanitizeForEcho(ref)}`,
      });
    }
  }
}

function checkReferences(skillDir: string, body: string, issues: Issue[]): void {
  const skillRoot = resolve(skillDir);

  const mdRefs = [...body.matchAll(MD_LINK_RE)]
    .map((m) => m[1]!)
    .filter((ref) => !/^(https?:\/\/|#|\/)/.test(ref));
  checkRefsExist(skillDir, skillRoot, mdRefs, "validate.reference.broken", "file", issues);

  const scriptRefs = [...body.matchAll(SCRIPT_REF_RE)].map((m) => m[1]!);
  checkRefsExist(skillDir, skillRoot, scriptRefs, "validate.script.broken", "script", issues);
}

interface Summary {
  fail: number;
  warn: number;
  ok: boolean;
}

function summarize(issues: Issue[]): Summary {
  const fail = issues.filter((i) => i.severity === "fail").length;
  const warn = issues.filter((i) => i.severity === "warn").length;
  return { fail, warn, ok: fail === 0 };
}

function emitText(issues: Issue[], summary: Summary): void {
  for (const issue of issues) {
    console.log(issueLine(issue));
  }
  console.log(`\n${summary.fail} fail(s), ${summary.warn} warning(s)`);
}

function emitJson(skillDir: string, issues: Issue[], summary: Summary): void {
  console.log(JSON.stringify({ skill_dir: skillDir, issues, summary }, null, 2));
}

const HELP = `Validate an agent skill against the Agent Skills specification.

Usage:
  validate_skill.ts <skill-directory> [--json] [--exit-on-warn]

Examples:
  validate_skill.ts path/to/my-skill
  validate_skill.ts ./skill --json
  validate_skill.ts ./skill --exit-on-warn

Options:
  --format <json|text>  Output format (default: text)
  --json                 Alias for --format json
  --quiet                Suppress informational stderr
  --exit-on-warn         Return non-zero exit if any WARN-level issues are found
  --help                 Show this help message
`;

export function main(argv: string[]): number {
  const { values, positionals } = parseArgs({
    args: argv,
    allowPositionals: true,
    options: {
      format: { type: "string", default: "text" },
      json: { type: "boolean", default: false },
      quiet: { type: "boolean", default: false },
      "exit-on-warn": { type: "boolean", default: false },
      help: { type: "boolean", default: false },
    },
  });

  if (values.help) {
    console.log(HELP);
    return 0;
  }

  const useJson = values.json || values.format === "json";
  const skillDirArg = positionals[0];
  if (!skillDirArg) {
    console.error(HELP);
    return 2;
  }

  const skillDir = resolve(skillDirArg.replace(/^~/, process.env.HOME ?? "~"));
  if (!existsSync(skillDir)) {
    emitError("validate_skill", `path does not exist: ${skillDir}`, {
      code: "validate.input.not-found",
      hint: "Check the path and try again.",
    });
    return 2;
  }
  if (!statSync(skillDir).isDirectory()) {
    emitError("validate_skill", `not a directory: ${skillDir}`, {
      code: "validate.input.not-dir",
      hint: "Argument must be a skill directory, not a file.",
    });
    return 2;
  }

  const issues = validate(skillDir);
  const summary = summarize(issues);
  if (useJson) {
    emitJson(skillDir, issues, summary);
  } else {
    emitText(issues, summary);
  }

  if (summary.fail) return 1;
  if (values["exit-on-warn"] && summary.warn) return 1;
  return 0;
}

const isMain = process.argv[1] && import.meta.url === `file://${process.argv[1]}`;
if (isMain) {
  process.exit(main(process.argv.slice(2)));
}

File: skills/skill-audit/scripts/validate_skill.py
#!/usr/bin/env python3
"""Validate an agent skill against the Agent Skills specification.

Usage:
    validate_skill.py <skill-directory> [--json] [--exit-on-warn]

Default output is one issue per line, prefixed with FAIL: or WARN: and a
machine-readable code in brackets — backwards compatible with prior behavior.
With --json, emits a structured object on stdout.

Exit codes:
    0   no FAIL-level issues (and no WARN if --exit-on-warn)
    1   one or more FAIL-level issues
    2   bad invocation (path missing, not a directory, etc.)
"""

from __future__ import annotations

import argparse
from dataclasses import asdict, dataclass
import json
from pathlib import Path
import re
import sys
from typing import Any

from skill_lib import emit_error, parse_frontmatter, sanitize_for_echo

NAME_RE = re.compile(r"^[a-z0-9]+(-[a-z0-9]+)*$")
_MD_LINK_RE = re.compile(r"\]\(([^)]+\.md)\)")
_SCRIPT_REF_RE = re.compile(r"`(scripts/[^`\s]+\.(?:py|sh|js|ts))`")
MAX_NAME = 64
MAX_DESCRIPTION = 1024
MIN_DESCRIPTION_WARN = 60
MAX_COMPATIBILITY = 500
ALLOWED_KEYS = frozenset(
    {
        # agentskills.io base spec
        "name",
        "description",
        "license",
        "compatibility",
        "metadata",
        "allowed-tools",
        # Platform extensions (Claude Code, Copilot, Codex, VS Code)
        "when_to_use",
        "argument-hint",
        "arguments",
        "disable-model-invocation",
        "user-invocable",
        "model",
        "effort",
        "context",
        "agent",
        "hooks",
        "paths",
        "shell",
    }
)


@dataclass
class Issue:
    severity: str  # "fail" or "warn"
    code: str
    message: str
    field: str | None = None

    def to_line(self) -> str:
        prefix = "FAIL" if self.severity == "fail" else "WARN"
        return f"{prefix}: [{self.code}] {self.message}"


def validate(skill_dir: Path) -> list[Issue]:
    issues: list[Issue] = []
    skill_md = skill_dir / "SKILL.md"
    if not skill_md.exists():
        return [
            Issue(
                "fail",
                "validate.skill-md.missing",
                f"SKILL.md not found at {sanitize_for_echo(skill_md)}",
            )
        ]

    text = skill_md.read_text(encoding="utf-8")
    fm, body = parse_frontmatter(text)
    if fm is None:
        return [
            Issue(
                "fail",
                "validate.frontmatter.missing",
                "SKILL.md is missing YAML frontmatter (must start with '---')",
            )
        ]

    _check_name(fm, skill_dir, issues)
    _check_description(fm, issues)
    _check_compatibility(fm, issues)
    _check_unknown_keys(fm, issues)
    _check_references(skill_dir, body, issues)
    return issues


def _check_name(fm: dict[str, Any], skill_dir: Path, issues: list[Issue]) -> None:
    name = fm.get("name", "")
    if not name:
        issues.append(Issue("fail", "validate.name.missing", "'name' field is required", "name"))
        return
    safe = sanitize_for_echo(name, max_len=80)
    if len(name) > MAX_NAME:
        issues.append(
            Issue(
                "fail",
                "validate.name.too-long",
                f"'name' is {len(name)} chars (max {MAX_NAME})",
                "name",
            )
        )
    if not NAME_RE.match(name):
        issues.append(
            Issue(
                "fail",
                "validate.name.bad-format",
                f"'name' must be lowercase a-z/0-9 with single hyphens, no leading/"
                f"trailing/consecutive hyphens. Got: {safe!r}",
                "name",
            )
        )
    if name != skill_dir.name:
        issues.append(
            Issue(
                "fail",
                "validate.name.dir-mismatch",
                f"'name' ({safe!r}) must match parent directory "
                f"({sanitize_for_echo(skill_dir.name)!r})",
                "name",
            )
        )


def _check_description(fm: dict[str, Any], issues: list[Issue]) -> None:
    desc = fm.get("description", "")
    if not desc:
        issues.append(
            Issue(
                "fail",
                "validate.description.missing",
                "'description' is required and must be non-empty",
                "description",
            )
        )
        return
    if len(desc) > MAX_DESCRIPTION:
        issues.append(
            Issue(
                "fail",
                "validate.description.too-long",
                f"'description' is {len(desc)} chars (max {MAX_DESCRIPTION})",
                "description",
            )
        )
    if len(desc) < MIN_DESCRIPTION_WARN:
        issues.append(
            Issue(
                "warn",
                "validate.description.too-short",
                f"'description' is only {len(desc)} chars — likely too short to "
                f"convey when to trigger",
                "description",
            )
        )


def _check_compatibility(fm: dict[str, Any], issues: list[Issue]) -> None:
    compat = fm.get("compatibility")
    if isinstance(compat, str) and compat and len(compat) > MAX_COMPATIBILITY:
        issues.append(
            Issue(
                "fail",
                "validate.compatibility.too-long",
                f"'compatibility' is {len(compat)} chars (max {MAX_COMPATIBILITY})",
                "compatibility",
            )
        )


def _check_unknown_keys(fm: dict[str, Any], issues: list[Issue]) -> None:
    unknown = set(fm.keys()) - ALLOWED_KEYS
    for key in sorted(unknown):
        safe = sanitize_for_echo(key, max_len=64)
        issues.append(
            Issue(
                "warn",
                "validate.frontmatter.unknown-key",
                f"unknown frontmatter key {safe!r} (spec defines: "
                f"{', '.join(sorted(ALLOWED_KEYS))})",
                key,
            )
        )


def _check_references(skill_dir: Path, body: str, issues: list[Issue]) -> None:
    skill_root = skill_dir.resolve()

    def _within_skill(target: Path) -> bool:
        try:
            target.relative_to(skill_root)
        except ValueError:
            return False
        return True

    for ref in _MD_LINK_RE.findall(body):
        if ref.startswith(("http://", "https://", "#", "/")):
            continue
        # Skip template placeholder paths like `references/<topic>.md`.
        if any(c in ref for c in "<>${}"):
            continue
        target = (skill_dir / ref).resolve()
        # Refuse to probe paths that escape the skill directory.
        if not _within_skill(target):
            continue
        if not target.exists():
            issues.append(
                Issue(
                    "warn",
                    "validate.reference.broken",
                    f"referenced file does not exist: {sanitize_for_echo(ref)}",
                )
            )

    for script_match in _SCRIPT_REF_RE.finditer(body):
        ref = script_match.group(1)
        # Skip placeholder paths like `scripts/<name>.py` or `scripts/${X}.py`.
        if any(c in ref for c in "<>${}"):
            continue
        target = (skill_dir / ref).resolve()
        if not _within_skill(target):
            continue
        if not target.exists():
            issues.append(
                Issue(
                    "warn",
                    "validate.script.broken",
                    f"referenced script does not exist: {sanitize_for_echo(ref)}",
                )
            )


def _summary(issues: list[Issue]) -> dict[str, Any]:
    fails = sum(1 for i in issues if i.severity == "fail")
    warns = sum(1 for i in issues if i.severity == "warn")
    return {"fail": fails, "warn": warns, "ok": fails == 0}


def _emit_text(issues: list[Issue], summary: dict[str, Any]) -> None:
    for issue in issues:
        print(issue.to_line())
    print(f"\n{summary['fail']} fail(s), {summary['warn']} warning(s)")


def _emit_json(skill_dir: Path, issues: list[Issue], summary: dict[str, Any]) -> None:
    payload = {
        "skill_dir": str(skill_dir),
        "issues": [asdict(i) for i in issues],
        "summary": summary,
    }
    print(json.dumps(payload, indent=2))


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(
        description="Validate an agent skill against the Agent Skills specification.",
        epilog="Examples:\n"
        "  validate_skill.py path/to/my-skill\n"
        "  validate_skill.py ./skill --json\n",
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    parser.add_argument("skill_dir", help="Path to the skill directory")
    parser.add_argument("--format", choices=["json", "text"], default="text", help="Output format.")
    parser.add_argument(
        "--json", dest="as_json", action="store_true", help="Alias for --format json."
    )
    parser.add_argument("--quiet", action="store_true", help="Suppress informational stderr.")
    parser.add_argument(
        "--exit-on-warn",
        action="store_true",
        help="Return non-zero exit if any WARN-level issues are found.",
    )
    args = parser.parse_args(argv)
    use_json = args.as_json or args.format == "json"

    skill_dir = Path(args.skill_dir).expanduser().resolve()
    if not skill_dir.exists():
        emit_error(
            "validate_skill",
            f"path does not exist: {skill_dir}",
            code="validate.input.not-found",
            hint="Check the path and try again.",
        )
        return 2
    if not skill_dir.is_dir():
        emit_error(
            "validate_skill",
            f"not a directory: {skill_dir}",
            code="validate.input.not-dir",
            hint="Argument must be a skill directory, not a file.",
        )
        return 2

    issues = validate(skill_dir)
    summary = _summary(issues)
    if use_json:
        _emit_json(skill_dir, issues, summary)
    else:
        _emit_text(issues, summary)

    if summary["fail"]:
        return 1
    if args.exit_on_warn and summary["warn"]:
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())
